### PR TITLE
hellofresh: use map instead of object

### PR DESCRIPTION
**Pull request by PepijnSenders**
## Summary

This PR refactors the datafile maps to use Map() instead of object. The datafile at HelloFresh is quite sizeable and doing all these object lookups and remappings are bad for performance. Map is up to 5 times faster in big objects!

### CODE REVIEW

**Comment by zashraf1985** [comment]
Thanks for the contribution @PepijnSenders !! There are many unit tests failing. While we review, if you have time can you take a look at why unit tests are failing.

**Comment by PepijnSenders** [comment]
@zashraf1985 yea I figured, if you think it's a good change I can go ahead and fix those tests.

**Comment by zashraf1985** [comment]
> @zashraf1985 yea I figured, if you think it's a good change I can go ahead and fix those tests.

It is a good change for sure. We developed Javascript SDK a few years ago in plain common JS. We gradually added bundling and typescript and are still in the process of gradually upgrading and optimizing. This is an ongoing process and there is a lot of room for optimization. We always appreciate and welcome contributions like this.

**Comment by PepijnSenders** [comment]
btw @zashraf1985 some of these tests are failing not because of my PR but because of your CI

**Comment by PepijnSenders** [comment]
@zashraf1985 all tests are fixed now :), would have been amazing if your specs were written in TS lol

**Comment by mikechu-optimizely** [comment]
Our test suite included some older versions of browsers that we're swapping out.  We're upgrading that suite now via another PR.  We'll then update this branch and merge it afterwards.  

Local testing looks good so far.

Thanks for hanging in there with us.

@PepijnSenders Can you sign our [Contributor License Agreement (CLA)](https://docs.google.com/a/optimizely.com/forms/d/e/1FAIpQLSf9cbouWptIpMgukAKZZOIAhafvjFCV8hS00XJLWQnWDFtwtA/viewform) when you have a moment?

**Review by zashraf1985 — CHANGES_REQUESTED** [review]
Thanks again for the contribution. I had one pass through it today and added some comments . It will be great if you can find some time to take a look at those. Can you also update your branch with the latest master.

File: packages/optimizely-sdk/lib/core/decision_service/index.ts [hunk 1]
```diff
@@ -451,15 +402,17 @@ export class DecisionService {
...
       } else {
         this.logger.log(
           LOG_LEVEL.INFO,
           LOG_MESSAGES.SAVED_VARIATION_NOT_FOUND,
-          MODULE_NAME, userId,
+          MODULE_NAME,
+          userId,
           variationId,
-          experiment.key,
+          experiment.key
```
[the remark this answers is not in the document]
Comment (PepijnSenders) [reply packages/optimizely-sdk/lib/core/decision_service/index.ts:415]: What about adding prettier to your setup? 😸

File: packages/optimizely-sdk/lib/core/decision_service/index.ts [hunk 2]
```diff
@@ -653,9 +591,9 @@ export class DecisionService {
   private getVariationForRollout(
     configObj: ProjectConfig,
     feature: FeatureFlag,
-    user: OptimizelyUserContext,
+    user: OptimizelyUserContext
```
[the remark this answers is not in the document]
Comment (PepijnSenders) [reply packages/optimizely-sdk/lib/core/decision_service/index.ts:594]: @zashraf1985 this is not according to your prettier configuration, I formatted all the files.

[the remark this answers is not in the document]
Comment (PepijnSenders) [reply packages/optimizely-sdk/lib/core/decision_service/index.ts:594]: @zashraf1985 this is not the case according to this prettier file: https://github.com/optimizely/javascript-sdk/blob/master/packages/optimizely-sdk/.prettierrc#L7

File: packages/optimizely-sdk/lib/core/decision_service/index.ts [hunk 3]
```diff
@@ -213,43 +202,23 @@ export class DecisionService {
     decideReasons.push(...decisionVariation.reasons);
     const variationId = decisionVariation.result;
     if (variationId) {
-      variation = configObj.variationIdMap[variationId];
+      const configVariation = configObj.variationIdMap.get(variationId);
+
+      if (configVariation) {
+        variation = configVariation;
+      }
```
[the remark this answers is not in the document]
Comment (PepijnSenders) [reply packages/optimizely-sdk/lib/core/decision_service/index.ts:209]: `.get()` might return `undefined`, so you want to make sure it exists before you set it.

**Comment by opti-jnguyen** [comment]
Merged recent changes from javascript-sdk master branch into this PR, however some issues remain regarding:

- [ ] `audiencesById.get` failing in `optimizely-sdk/lib/core/audience_evaluator/index.ts`
- [ ] New ODP segments feature test failing for test "should contain all expected unique odp segments in allSegments"

Will need to fix these and ensure `npm run test` succeeds before proceeding.

**Comment by opti-jnguyen** [comment]
@PepijnSenders - thank you for the contribution! After internal testing there's a couple issues that are preventing us from merging at this time:

1. Testing usage of the Map implementation on very large datafile sizes (>2 MB) resulted in slower runtimes of >=10% on average vs. the original object implementation. Methods of testing included manual testing with `performance.now()` and testing leveraging existing unit tests.

2. General risk due to changing the original implementation without much ancillary benefit.

---

Going forward we'll be making an attempt at experimenting between using Map and object using your implementation as inspiration and benchmarking differences as we go and implement according based on the results we see.

For the time being though, we'll be closing this PR but feel free to raise an issue if you have any comments, questions, or found different benchmarks when testing yourself. Thank you!

**Pull request closed without merging**